src/code/z_horse.c

```diff
@@ -1,21 +1,233 @@
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F3B2C.s")
+        if (actor->params == BG_UMAJUMP_TYPE_2) {
+            return ((BgUmajump*)actor)->objectIndex;
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F3B68.s")
+    return -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F3C44.s")
+// unused
+s32 func_800F39B4(PlayState* play, s32 pathIndex, s32 pointIndex, Vec3s* dst, s16* arg4) {
+    Path* path = &play->setupPathList[pathIndex];
+    Vec3s* points;
+    s32 count = path->count;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F3ED4.s")
+    dst->x = 0;
+    dst->y = 0;
+    dst->z = 0;
+    *arg4 = 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F40A0.s")
+    if (count == 0) {
+        return false;
+    }
+    if (pointIndex >= count) {
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F415C.s")
+    points = Lib_SegmentedToVirtual(path->points);
+    points += pointIndex;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_horse/func_800F41E4.s")
+    dst->x = points->x;
+    dst->y = points->y;
+    dst->z = points->z;
+    *arg4 = 0;
+
+    return true;
+}
+
+typedef struct struct_801BDAA8 {
+    /* 0x0 */ s16 sceneId;
+    /* 0x2 */ s16 sceneLayerMinusOne;
+} struct_801BDAA8; // size = 0x4
+
+struct_801BDAA8 sHorseValidScenes[] = {
+    { SCENE_00KEIKOKU, 0 },      // Termina Field
+    { SCENE_24KEMONOMITI, 0 },   // Road to southern swap
+    { SCENE_F01, 0 },            // Romani ranch
+    { SCENE_KOEPONARACE, 0 },    // Gorman track
+    { SCENE_20SICHITAI, 0 },     // Southern swamp poisoned
+    { SCENE_20SICHITAI2, 0 },    // Souther swamp clear
+    { SCENE_30GYOSON, 0 },       // Great bay coast
+    { SCENE_31MISAKI, 0 },       // Zora cape
+    { SCENE_ROMANYMAE, 0 },      // Milk road
+    { SCENE_IKANAMAE, 0 },       // Road to Ikana
+    { SCENE_13HUBUKINOMITI, 0 }, // Path to Mountain Village
+};
+
+s32 Horse_IsValidSpawn(s16 sceneId) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(sHorseValidScenes); i++) {
+        if (sceneId == sHorseValidScenes[i].sceneId) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void Horse_ResetHorseData(PlayState* play) {
+    gSaveContext.save.saveInfo.horseData.sceneId = SCENE_F01;
+    gSaveContext.save.saveInfo.horseData.pos.x = -1420;
+    gSaveContext.save.saveInfo.horseData.pos.y = 257;
+    gSaveContext.save.saveInfo.horseData.pos.z = -1285;
+    gSaveContext.save.saveInfo.horseData.yaw = 0x2AAA;
+}
+
+s32 gHorseIsMounted = false;
+s32 D_801BDAA0 = false;
+s32 gHorsePlayedEponasSong = false;
+
+struct_801BDAA8 D_801BDAA8[] = {
+    { SCENE_00KEIKOKU, 5 - 1 },      // Termina Field - First Cycle
+    { SCENE_30GYOSON, 1 - 1 },       // Great Bay Coast - Post-Gyorg
+    { SCENE_31MISAKI, 1 - 1 },       // Zora Cape - Post-Gyorg
+    { SCENE_13HUBUKINOMITI, 1 - 1 }, // Path to Mountain Village - Post-Goht
```

Comment:
Afaik, layers above 0 are generally meant for cutscenes (but not always) so Epona generally cannot spawn in layers above 0. These overwrite that rule, since these aren't actually cutscene layers, but natural playable layers.